### ISSUE TITLE
[export] Cannot view a tensor with shape torch.Size([1, 512, 32, 128]) and strides (2097152, 128, 65536, 1) as a tensor with shape (1, 512, 4096)

### ISSUE THREAD

**Issue by justinchuby** [pytorch/pytorch#136543]
# Notes

The errors is happening in `ExportedProgram.run_decompositions()` call: message is `Cannot view a tensor with shape torch.Size([1, 512, 32, 128]) and strides (2097152, 128, 65536, 1) as a tensor with shape (1, 512, 4096)`.

@tugsbayasgalan do you know if anything about the `aten.view.default` ref or if it has any issues?

https://github.com/pytorch/pytorch/blob/ebfcbe0822caaf361a8f487a77db786a58543acf/torch/_refs/__init__.py#L4627-L4633

# PyTorch ONNX Conversion Report

```
✅ Obtain model graph with `torch.export.export`
⚪ Obtain model graph with `torch.export.export(..., strict=False)`
⚪ Obtain model graph with `torch.jit.trace`
❌ Translate the graph into ONNX
⚪ Run `onnx.checker` on the ONNX model
⚪ Execute the model with ONNX Runtime
⚪ Validate model output accuracy
```

## Error messages

```pytb


Traceback (most recent call last):

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/onnx/_internal/exporter/_core.py", line 1089, in export
    decomposed_program = _prepare_exported_program_for_export(

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/onnx/_internal/exporter/_core.py", line 727, in _prepare_exported_program_for_export
    exported_program = _fx_passes.decompose_with_registry(exported_program, registry)

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/onnx/_internal/exporter/_fx_passes.py", line 20, in decompose_with_registry
    return exported_program.run_decompositions(decomp_table)

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/export/exported_program.py", line 114, in wrapper
    return fn(*args, **kwargs)

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/export/exported_program.py", line 1080, in run_decompositions
    return _decompose_exported_program(

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/export/exported_program.py", line 628, in _decompose_exported_program
    gm, new_graph_signature = _decompose_and_get_gm_with_new_signature_constants(

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/export/exported_program.py", line 421, in _decompose_and_get_gm_with_new_signature_constants
    gm, graph_signature = aot_export_module(

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/_functorch/aot_autograd.py", line 1246, in aot_export_module
    fx_g, metadata, in_spec, out_spec = _aot_export_function(

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/_functorch/aot_autograd.py", line 1480, in _aot_export_function
    fx_g, meta = create_aot_dispatcher_function(

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/_functorch/aot_autograd.py", line 522, in create_aot_dispatcher_function
    return _create_aot_dispatcher_function(

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/_functorch/aot_autograd.py", line 623, in _create_aot_dispatcher_function
    fw_metadata = run_functionalized_fw_and_collect_metadata(

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/_functorch/_aot_autograd/collect_metadata_analysis.py", line 173, in inner
    flat_f_outs = f(*flat_f_args)

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/_functorch/_aot_autograd/utils.py", line 182, in flat_fn
    tree_out = fn(*args, **kwargs)

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/_functorch/_aot_autograd/traced_function_transforms.py", line 859, in functional_call
    out = PropagateUnbackedSymInts(mod).run(

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/fx/interpreter.py", line 146, in run
    self.env[node] = self.run_node(node)

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/fx/experimental/symbolic_shapes.py", line 5619, in run_node
    result = super().run_node(n)

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/fx/interpreter.py", line 203, in run_node
    return getattr(self, n.op)(n.target, args, kwargs)

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/fx/interpreter.py", line 275, in call_function
    return target(*args, **kwargs)

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/_ops.py", line 716, in __call__
    return self._op(*args, **kwargs)

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/_subclasses/functional_tensor.py", line 535, in __torch_dispatch__
    outs_unwrapped = func._op_dk(

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/utils/_stats.py", line 21, in wrapper
    return fn(*args, **kwargs)

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/_subclasses/fake_tensor.py", line 1238, in __torch_dispatch__
    return self.dispatch(func, types, args, kwargs)

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/_subclasses/fake_tensor.py", line 1692, in dispatch
    return self._cached_dispatch_impl(func, types, args, kwargs)

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/_subclasses/fake_tensor.py", line 1339, in _cached_dispatch_impl
    output = self._dispatch_impl(func, types, args, kwargs)

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/_subclasses/fake_tensor.py", line 2039, in _dispatch_impl
    r = func(*args, **kwargs)

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/_ops.py", line 716, in __call__
    return self._op(*args, **kwargs)

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/_refs/__init__.py", line 4592, in view
    return _reshape_view_helper(a, *shape, allow_copy=False)

  File "/home/xadupre/.local/lib/python3.9/site-packages/torch/_refs/__init__.py", line 3755, in _reshape_view_helper
    raise ValueError(msg)

ValueError: Cannot view a tensor with shape torch.Size([1, 512, 32, 128]) and strides (2097152, 128, 65536, 1) as a tensor with shape (1, 512, 4096)!

While executing %view_4 : [num_users=1] = call_function[target=torch.ops.aten.view.default](args = (%transpose_4, [1, 512, -1]), kwargs = {})
Original traceback:
  File "/home/xadupre/github/private_occ/bash_bench/_bash_bench_model_runner.py", line 204, in __call__
    return self.model(*args, **kwargs)
  File "/home/xadupre/.local/lib/python3.9/site-packages/transformers/models/llama/modeling_llama.py", line 1189, in forward
    outputs = self.model(
  File "/home/xadupre/.local/lib/python3.9/site-packages/transformers/models/llama/modeling_llama.py", line 1001, in forward
    layer_outputs = decoder_layer(
  File "/home/xadupre/.local/lib/python3.9/site-packages/transformers/models/llama/modeling_llama.py", line 734, in forward
    hidden_states, self_attn_weights, present_key_value = self.self_attn(
  File "/home/xadupre/.local/lib/python3.9/site-packages/transformers/models/llama/modeling_llama.py", line 670, in forward
    attn_output = attn_output.view(bsz, q_len, -1)


```

## Exported program

```python
...
 # File: /home/xadupre/.local/lib/python3.9/site-packages/transformers/models/llama/modeling_llama.py:654 in forward, code: value_states = value_states.contiguous()
clone_2: "f16[1, 32, 512, 128]" = torch.ops.aten.clone.default(transpose_3, memory_format = torch.contiguous_format)

 # File: /home/xadupre/.local/lib/python3.9/site-packages/transformers/models/llama/modeling_llama.py:660 in forward, code: attn_output = torch.nn.functional.scaled_dot_product_attention(
unsqueeze_8: "f16[1, 512, 513]" = torch.ops.aten.unsqueeze.default(mul, 0)
unsqueeze_9: "f16[1, 1, 512, 513]" = torch.ops.aten.unsqueeze.default(unsqueeze_8, 1);  unsqueeze_8 = None
slice_14: "f16[1, 1, 512, 513]" = torch.ops.aten.slice.Tensor(unsqueeze_9, 2, 0, 9223372036854775807);  unsqueeze_9 = None
slice_15: "f16[1, 1, 512, 513]" = torch.ops.aten.slice.Tensor(slice_14, 3, 0, 9223372036854775807);  slice_14 = None
expand_2: "f16[1, 1, 512, 513]" = torch.ops.aten.expand.default(slice_15, [1, 1, -1, -1]);  slice_15 = None
slice_16: "f16[1, 1, 512, 513]" = torch.ops.aten.slice.Tensor(expand_2, 0, 0, 9223372036854775807);  expand_2 = None
slice_17: "f16[1, 1, 512, 513]" = torch.ops.aten.slice.Tensor(slice_16, 1, 0, 9223372036854775807);  slice_16 = None
slice_18: "f16[1, 1, 512, 513]" = torch.ops.aten.slice.Tensor(slice_17, 2, 0, 9223372036854775807);  slice_17 = None
slice_19: "f16[1, 1, 512, 512]" = torch.ops.aten.slice.Tensor(slice_18, 3, 0, 512);  slice_18 = None
scaled_dot_product_attention: "f16[1, 32, 512, 128]" = torch.ops.aten.scaled_dot_product_attention.default(clone, clone_1, clone_2, slice_19);  clone = clone_1 = clone_2 = slice_19 = None

 # File: /home/xadupre/.local/lib/python3.9/site-packages/transformers/models/llama/modeling_llama.py:669 in forward, code: attn_output = attn_output.transpose(1, 2).contiguous()
transpose_4: "f16[1, 512, 32, 128]" = torch.ops.aten.transpose.int(scaled_dot_product_attention, 1, 2);  scaled_dot_product_attention = None

 # File: /home/xadupre/.local/lib/python3.9/site-packages/transformers/models/llama/modeling_llama.py:670 in forward, code: attn_output = attn_output.view(bsz, q_len, -1)
view_4: "f16[1, 512, 4096]" = torch.ops.aten.view.default(transpose_4, [1, 512, -1]);  transpose_4 = None

 # File: /home/xadupre/.local/lib/python3.9/site-packages/transformers/models/llama/modeling_llama.py:672 in forward, code: attn_output = self.o_proj(attn_output)
linear_3: "f16[1, 512, 4096]" = torch.ops.aten.linear.default(view_4, p_model_model_layers_0_self_attn_o_proj_weight);  view_4 = p_model_model_layers_0_self_attn_o_proj_weight = None
...
```

## Analysis

PyTorch ONNX Conversion Analysis

## Model Information

The model has 6738546688 parameters and 64 buffers (non-trainable parameters).
Number of parameters per dtype:
```python
defaultdict(<class 'int'>, {torch.float16: 6738546688})
```
Number of buffers per dtype:
```python
defaultdict(<class 'int'>, {torch.float16: 64})
```

Of the call_function nodes, the counts of operators used are:

- `aten.slice.Tensor`: 323
- `aten.mul.Tensor`: 293
- `aten.linear.default`: 225
- `aten.add.Tensor`: 193
- `aten._to_copy.default`: 135
- `aten.unsqueeze.default`: 132
- `aten.view.default`: 129
- `aten.transpose.int`: 128
- `aten.clone.default`: 96
- `aten.pow.Tensor_Scalar`: 65
- `aten.mean.dim`: 65
- `aten.rsqrt.default`: 65
- `aten.neg.default`: 64
- `aten.cat.default`: 64
- `aten.expand.default`: 33
- `aten.scaled_dot_product_attention.default`: 32
- `aten.silu.default`: 32
- `<built-in function getitem>`: 2
- `aten.embedding.default`: 1
- `aten.arange.start`: 1
- `aten.full.default`: 1
- `aten.triu.default`: 1
- `aten.arange.default`: 1
- `aten.gt.Tensor`: 1
- `wrap_with_autocast`: 1

## ONNX Conversion Information

The model contains operators the dispatcher could not find registered ONNX decompositions for. This may be due to missing implementations, decompositions not registered correctly, or a bug in the dispatcher.

Errors grouped by operator:

- `wrap_with_autocast`:     No decompositions registered for the real-valued input. Example node: `%wrap_with_autocast : [num_users=2] = call_function[target=torch.ops.higher_order.wrap_with_autocast](args = (cuda, None, False, None, %submod_3, %expand_1, %_to_copy_1), kwargs = {})`. All nodes: `[wrap_with_autocast]`

## Profiling result

```

  _     ._   __/__   _ _  _  _ _/_   Recorded: 16:08:25  Samples:  24301
 /_//_/// /_\ / //_// / //_'/ //     Duration: 57.754    CPU time: 57.757
/   _/                      v4.7.3

Profile at /home/xadupre/.local/lib/python3.9/site-packages/torch/onnx/_internal/exporter/_core.py:646

57.753 export  torch/onnx/_internal/exporter/_core.py:932
├─ 36.167 save  torch/export/__init__.py:377
│     [60 frames hidden]  torch, <built-in>, copy, <string>, zi...
└─ 20.468 TorchExportStrategy.__call__  torch/onnx/_internal/exporter/_capture_strategies.py:79
      [215 frames hidden]  torch, <string>
         8.183 _fn  torch/_dynamo/eval_frame.py:431
         └─ 1.191 WrappedModelBase.__call__  bash_bench/_bash_bench_model_runner.py:203
            └─ 1.190 _fn  torch/_dynamo/eval_frame.py:628
                  [11 frames hidden]  torch, <eval_with_key>

```

cc @ezyang @chauhang @penguinwu @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4

**Comment by justinchuby** [comment]
cc @xadupre @gramalingam

**Comment by malfet** [comment]
Removing triaged and onnx labels, @guangy10 can you please reply to this one?

**Comment by pianpwk** [comment]
@justinchuby could you retry on latest main? It might have been fixed by https://github.com/pytorch/pytorch/pull/136492

**Comment by justinchuby** [comment]
@xadupre Could you comment on if the issue persists?

**Comment by xadupre** [comment]
The issue persists but one function decomposition scaled_dot_product_attention is wrong. I modified the code of ``run_decompositions`` (https://github.com/pytorch/pytorch/blob/main/torch/export/exported_program.py#L1095) to remove them and it now works. So the bug is in one decomposition function in ``cia_to_decomp`` whose name contains ``scaled_dot``.

```python
        (
            cia_to_decomp,
            python_decomp_table,
        ) = _split_decomp_table_to_cia_and_python_decomp(_decomp_table)

        new_table = {}
        for k, v in cia_to_decomp.items():
            if "scaled_dot" not in k.name():
                new_table[k] = v
        cia_to_decomp = new_table

        return _decompose_exported_program(
            self,
            cia_to_decomp=cia_to_decomp,
            python_decomp_table=python_decomp_table,
            joint_loss_index=None,
        )
```

**Comment by justinchuby** [comment]
@pianpwk @tugsbayasgalan

**Comment by tugsbayasgalan** [comment]
Is there a repro we can try out? @xadupre Can you clarify what the wrong decomposition behavior you observed in the scaled_dot?

**Comment by xadupre** [comment]
Here is a simple test based I made based on the model which fails. The issue does not seem to come from ``scaled_dot_product_attention`` but from ``view``.

```python
    def test_scaled_dot_product_attention(self):

        class DummyModel(torch.nn.Module):
            def __init__(self, enable_math: bool):
                super().__init__()
                self.enable_math = False

            def forward(self, query, key, value):
                res = torch.nn.functional.scaled_dot_product_attention(
                    query, key, value
                )
                rest = res.transpose(0, 1)
                final = rest.view(8, 32, 128 * 64)
                return final

        model = DummyModel(False)
        device = "cpu"

        query = torch.rand(32, 8, 128, 64, dtype=torch.float16, device=device)
        key = torch.rand(32, 8, 128, 64, dtype=torch.float16, device=device)
        value = torch.rand(32, 8, 128, 64, dtype=torch.float16, device=device)
        expected = model(query, key, value)
        self.assertEqual(expected.dtype, torch.float16)
        self.assertEqual(expected.shape, (8, 32, 8192))

        cpl = torch.compile(model)
        new_output = cpl(query, key, value)
        self.assertEqual(new_output.dtype, torch.float16)
        self.assertEqual(new_output.shape, (8, 32, 8192))
        self.assertEqualArray(expected, new_output)        

        export = torch.export.export(model, (query, key, value))
        # Next line fails here due to
        # Cannot view a tensor with shape torch.Size([8, 32, 128, 64]) and strides
        # (64, 512, 16384, 1) as a tensor with shape (8, 32, 8192)
        export.run_decompositions()
```

**Comment by tugsbayasgalan** [comment]
Hmm actually even after patching your change (which sort of reverts the new change introduced in export), I am getting the same error. Does it used to work and recently stopped working?

**Comment by justinchuby** [comment]
I think so. I remeber starting to observe this after the preserve_ops refactoring

**Comment by xadupre** [comment]
The patch I made was for a specific model. It does not fix the test I shared. I believe one decomposition applies on the sequence transpose + view and it does take into account the strides. Maybe if we run a decomposition with insert a clone between both, that would solve it.

**Comment by xadupre** [comment]
Did you add --remove_shape_info at the end of the command line?

I'll check again tomorrow.

Xavier

**Comment by tugsbayasgalan** [comment]
Is there a way to access the model code directly? But i think putting a clone would work.

**Comment by xadupre** [comment]
> Is there a way to access the model code directly? But i think putting a clone would work.

Sorry, I replied to another issue in the wrong thread. The example with the DummyModel is a standalone example. I believe if run_decompositions is fixed for this one, it should fix onnx dynamo converter as well.

**Comment by tugsbayasgalan** [comment]
Right! I would like to also understand why this issue started showing up. But i agree we need to fix this issue as well.

**Comment by sainueng** [comment]
I experienced the same issue, going to torch-mlir instead of onnx. I also had to apply the change to `run_decompositions`
Following is my repro, which downloads llama 2 off of Huggingface.

```
import torch
# print("\nInstalled PyTorch, version:", torch.__version__)

# Load model directly
from transformers import AutoTokenizer, AutoModelForCausalLM

tokenizer = AutoTokenizer.from_pretrained("meta-llama/Llama-2-7b-hf")
model = AutoModelForCausalLM.from_pretrained("meta-llama/Llama-2-7b-hf")

# This is just a simple demo to get some data flowing through the model.
# Depending on this model and what input it expects (text, image, audio, etc.)
# this might instead use a specific Processor class. For Whisper,
# WhisperProcessor runs audio input pre-processing and output post-processing.
example_prompt = "Hello world!"
example_encoding = tokenizer(example_prompt, return_tensors="pt")
example_input = example_encoding["input_ids"].cpu()
example_args = (example_input,)
print(example_encoding)
print(example_input)

exported_program = torch.export.export(model, example_args)

from torch_mlir import fx
m = fx.export_and_import(model, *example_args)
print(m)


```

**Comment by justinchuby** [comment]
Is it possible to bisect the pytorch build using the repro code?

**Comment by xadupre** [comment]
The bug disappears with PR #136543 but I run it inside the onnx exporter. The same trick can be applied for torch_mlir. That does not fix the root cause of the bug but it should unblock us until it is fixed.

**Comment by justinchuby** [comment]
I was running git bisect but realized even with `release/2.5` the pattern above fails. I am starting to think it's because of some transformers updates that triggered this bug.

Scripts I used to bisect:

```sh
#!/usr/bin/env bash

git submodule update --init --recursive
python setup.py develop
python bisectscript.py
```

bisectscript.py

```python
import torch

class Model(torch.nn.Module):

    def forward(self, query, key, value):
        res = torch.nn.functional.scaled_dot_product_attention(
            query, key, value
        )
        rest = res.transpose(0, 1)
        final = rest.view(8, 32, 128 * 64)
        return final

model = Model()
device = "cpu"

query = torch.rand(32, 8, 128, 64, dtype=torch.float16, device=device)
key = torch.rand(32, 8, 128, 64, dtype=torch.float16, device=device)
value = torch.rand(32, 8, 128, 64, dtype=torch.float16, device=device)

export = torch.export.export(model, (query, key, value))
# Next line fails here due to
# Cannot view a tensor with shape torch.Size([8, 32, 128, 64]) and strides
# (64, 512, 16384, 1) as a tensor with shape (8, 32, 8192)
export.run_decompositions()
```

**Comment by justinchuby** [comment]
Let me see if I can add a contiguous call somewhere

**Comment by justinchuby** [comment]
@tugsbayasgalan @angelayi bump. This is an export issue.

**Comment by luoxinkaikevin** [comment]
Hi~I also meet this error when export llava model to stablehlo mlir. Are there any new developments or solutions to this problem?
Thanks for your reply in advance!

**Comment by pianpwk** [comment]
I might be wrong, but I feel like it's an issue with this check in `_collapse_view_helper`: https://github.com/pytorch/pytorch/blob/eabef5000fc654e3f7ea186e8630130f05f46991/torch/_prims/__init__.py#L1417

It leads to an error being thrown when the resulting collapse doesn't have strides of a contiguous tensor, but the original input tensor wasn't contiguous in the first place.

**Comment by pianpwk** [comment]
I think the issue here is that the SDPA op outputs a contiguous layout when not decomposed, but a non-contiguous layout when decomposed, and so the use of the view op is only exposed when decomposed.

For example you'll see this contiguous output when running in eager:
```
(Pdb) torch.nn.functional.scaled_dot_product_attention(query, key, value).shape
torch.Size([32, 8, 128, 64])
(Pdb) torch.nn.functional.scaled_dot_product_attention(query, key, value).stride()
(65536, 8192, 64, 1)
```

But if I look at the decomposed version (i.e. export a model containing only the SDPA op + transpose & run decompositions on it), then I see this as the output:
```
class DummyModel(torch.nn.Module):
    def forward(self, query, key, value):
        res = torch.nn.functional.scaled_dot_product_attention(
            query, key, value
        )
        return res.transpose(0, 1)
ep = torch.export.export(model, (query, key, value))
ep = ep.run_decompositions()
output = ep.module()(query, key, value)
print(output.shape, output.stride())

# torch.Size([8, 32, 128, 64]) (64, 512, 16384, 1)
```

And a view op won't work on this, unless you call contiguous() on it, or use reshape() instead of view(). This breaks even in eager mode, if you try to call view on a non-contiguous tensor:
```
>>> x = torch.randn(1,2,3,4).permute(2,0,3,1).view(-1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: view size is not compatible with input tensor's size and stride (at least one dimension spans across two contiguous subspaces). Use .reshape(...) instead.
>>> torch.randn(1,2,3,4).permute(2,0,3,1).contiguous().view(-1).shape
torch.Size([24])
>>> torch.randn(1,2,3,4).permute(2,0,3,1).reshape(-1).shape
torch.Size([24])
```

So I think there's 2 options here:
1) you can change your view() call from `final = rest.view(8, 32, 128 * 64)` to `final = rest.contiguous().view(8, 32, 128 * 64)` - I've verified that this works with run_decompositions and the whole flow succeeds.
2) you can change your view() call from `final = rest.view(8, 32, 128 * 64)` to `final = rest.reshape(8, 32, 128 * 64)` - this should work, but I'm seeing that the reshape is decomposing to view on the 1st export call, which causes the same error in decompositions.. so this doesn't seem to work.

**Comment by justinchuby** [comment]
Potentially when decomposing the logic needs to recognize that the output is no longer contiguous and replace view with reshape (or add contiguous to the output) in the graph?

**Comment by tugsbayasgalan** [comment]
Can you try on latest master? This issue might be fixed now

**Comment by justinchuby** [comment]
Thanks - I think it is fixed. Do you know how it was fixed?
